### ISSUE TITLE
Offline mobile version/responsive issues

### ISSUE THREAD

**Issue by aashish108** [zadam/trilium#856]
Hi! II've been trying out Trilium on y Pi and it looks like a great app! 

Is there a way to get offline mode on mobile web Android pls? Been playing around with it. It doesn't have to be an Android app. I think you can save websites on Chrome (etc) and open it like an app?

Also been trying to use Trilium on a web panel on desktop (mobile width) and sometimes at full width on desktop at the same time (testing) and its not really responsive as it depends on login what width it is or if it has the ?mobile URL parameter, right?

**Comment by zadam** [comment]
Hello, mobile version is offline only. Once Android supports Electron (or the other way round!) we can make offline version as well :)

For the second question - you're right that this switching happens based on the login and not responsively. However in practice this should not be a problem since you can't really switch between phone and computer within one session ...

**Comment by gerroon** [comment]
@zadam do you know if this can be compiled on Android using something like Termux? That way one cna have a local copy and sync with another instance.

**Comment by zadam** [comment]
Did a quick google and seems like it should be possible - you can also run docker using Termux and there are already some ARM builds linked from here: https://github.com/zadam/trilium/wiki/Docker-server-installation

**Comment by gerroon** [comment]
@zadam thanks, I will see what I can do.

**Comment by aashish108** [comment]
I had a look at the Android Electron issue and, judging by the look of things, it will never be released due to too many issues. So I guess we will never get mobile app with offline storage!  :/

I guess its too simple to suggest HTML5 web storage? :O

**Comment by zadam** [comment]
Yes, the trouble is that HTML5 web storage and other solutions would require different codebase and that's simply not sustainable for me.

**Comment by capi** [comment]
I've just successfully compiled the server-side of Trilium under Termux on Android. It runs great. If we could add the possibility to setup the _web_ version as a sync _client_, I think this could be the basis for an offline-capable version for Android (for geeks).

**Comment by gerroon** [comment]
@capi how about distributing it under Termux? That would make it easy for alot of people.

**Comment by jhc86** [comment]
Thank you!

I tried to figure out Docker before, but I wasn't successful at the time. (and I'm forever struggling to decide what's best to spend time on) This additional information is really exciting and encouraging me to try it again.

I'd like to echo Sytten's enthusiasm for a way to donate to this project, (https://github.com/zadam/trilium/issues/826) but I definitely understand the concerns about legality.

**Comment by gerroon** [comment]
@ jhc86 if youa re referring to use this on Android locally, you do not need Docker, just install Termux and follow the build instructions from the main page, at least that is my guess.

**Comment by jhc86** [comment]
Here's what I found:

pkg install nodejs-lts gives segmentation fault, so I used pkg remove nodejs-lts and tried again with nodejs instead

Other packages I used: wget, git, python2, and hunspell

I tried both release 40.2 tar.gz (tar -xf releaseName.tar.gx) and git clone https://github.com/zadam/trilium

Both got hung up on: (fuzzy memory, phone died) spellcheck_use_hunspell
So I tried pkg install hunspell but still got error
npm install hunspell says no package found
I think I also tried npm install spellcheck and that failed but I can't remember exactly

I can try to erase termux storage and try again when I have another chance

**Comment by capi** [comment]
@gerroon It was just a first try, it's a long way until a package. :-)

@jhc86 for me the following setup worked, I just had one build-error for something related to PNG, which I used a "--ignore-scripts" for npm for this one package:

```
pkg install nodejs-lts python build-essential
git clone https://github.com/zadam/trilium.git
cd trilium
git checkout v0.40.3
# vim package.json # remove anything electron
npm install
TRILIUM_HOST=127.0.0.1 node src/www
```

The important part for me was to remove anything related to `electron` from the package.json.

**Comment by zadam** [comment]
> I've just successfully compiled the server-side of Trilium under Termux on Android. It runs great. If we could add the possibility to setup the web version as a sync client, I think this could be the basis for an offline-capable version for Android (for geeks).

Interesting! Web version can work as a sync client, so far it has been disabled since there wasn't any use case for that ...

**Comment by gerroon** [comment]
@capi thanks, could you also post you package.json?

**Comment by capi** [comment]
> Interesting! Web version can work as a sync client, so far it has been disabled since there wasn't any use case for that ...

@zadam Is there an easy way to patch the selection for the setup screen in? Just for testing? (I have not yet looked into the code, to be honest).

@gerroon attached is my package.json for **v0.40.3**: [package.json.zip](https://github.com/zadam/trilium/files/4188202/package.zip)
I had some issues with compiling `pngquant-bin`, I worked around that by explicitly installing it without running install scripts (no idea yet what that causes): `npm install pngquant-bin@5.0.2 --ignore-scripts`

So the whole sequence is:
```
pkg install nodejs-lts python build-essential
git clone https://github.com/zadam/trilium.git
cd trilium
# vim package.json # remove anything electron
npm install
npm install pngquant-bin@5.0.2 --ignore-scripts
npm install
node src/www
```

I connected to the server with my desktop browser, I did not yet test local Android Chrome. My first experiments show, that the server running works with most things I tried. But it is _really_ slow compared to running it on x86 hardware (e.g. upload of picture takes long).

**Comment by capi** [comment]
I've just compile v0.40.4 and used the sync support of the server instance. So far it works great!

I applied this little patch to get it running: [trilium-v0.40.4-termux.zip](https://github.com/zadam/trilium/files/4252149/trilium-v0.40.4-termux.zip)

It's still necessary to `npm install` pngquant-bin explicitly:
```
npm install
npm install pngquant-bin@5.0.2 --ignore-scripts
npm install
node src/www
```

**Comment by leoossa** [comment]
Have you guys tried Cordova?
https://medium.com/the-web-tub/electron-on-cordova-29ede5d6d789

**Comment by rogatty** [comment]
@capi thanks, works for me too. You can automate `# vim package.json # remove anything electron` with:
```
mv package.json package.json.backup
cat package.json.backup | grep -v electron > package.json

**Comment by tejasvi** [comment]
To anyone else stumbling across, I have created a decent [mobile wrapper](https://github.com/zadam/trilium/files/4806879/trilium.zip) for android. It can connect to both local and remote servers. Though the local server will provide offline access and lower latency.
First, install termux from the play store. Then install server directly from master or from [release](https://github.com/zadam/trilium/releases/) depending on the version of your other clients.
<pre>pkg install nodejs python build-essential --no-install-recommends -y
# Choose one:
# wget https://github.com/zadam/trilium/archive/v0.xx.x[-beta].zip
    # unzip v0.*.zip
    # cd trilium-0.*
# wget https://github.com/zadam/trilium/archive/stable.zip`
    # unzip stable.zip
    # cd trilium-stable
mv package-lock.json package-lock.json.bak
sed -i.bak '/electron/d' package.json
npm install pngquant-bin@5.0.2 --ignore-scripts
npm install
node src/www</pre>
Then extract [trilium.zip](https://github.com/zadam/trilium/files/4806879/trilium.zip) and install the apk. Use `http://localhost:8080` for default local server port.
Pro tip: Sync with the remote server if you have one by going to the desktop version of the browser.

![trilium](https://user-images.githubusercontent.com/45873379/85184629-08ce1a80-b2ae-11ea-95d4-8c91491e09ed.gif)

**Comment by tejasvi** [comment]
To anyone else stumbling across, I have created a decent [mobile wrapper](https://github.com/zadam/trilium/files/4806879/trilium.zip) for android. It can connect to both local and remote servers. Though the local server will provide offline access and lower latency.
First, install termux from the play store. Then install server directly from master or from [release](https://github.com/zadam/trilium/releases/) depending on the version of your other clients.
<pre>pkg install nodejs python build-essential --no-install-recommends -y
# Choose one:
# wget https://github.com/zadam/trilium/archive/v0.xx.x[-beta].zip
    # unzip v0.*.zip
    # cd trilium-0.*
# wget https://github.com/zadam/trilium/archive/stable.zip`
    # unzip stable.zip
    # cd trilium-stable
mv package-lock.json package-lock.json.bak
sed -i.bak '/electron/d' package.json
npm install pngquant-bin@5.0.2 --ignore-scripts
npm install
node src/www</pre>
Then extract [trilium.zip](https://github.com/zadam/trilium/files/4806879/trilium.zip) and install the apk. Use `http://localhost:8080` for default local server port.
Pro tip: Enable remote sync by going to the desktop version of the site.

![trilium](https://user-images.githubusercontent.com/45873379/85184629-08ce1a80-b2ae-11ea-95d4-8c91491e09ed.gif)

**Comment by capi** [comment]
@tejasvi Would you be willing to put up the wrapper on a Github repo under a comparable license to trilium itself? I think we could evolve that thing over the time into something very useful!

**Comment by ehopps** [comment]
@tejasvi Just came across your mobile wrapper and I'm seconding capi's comment above. Would you be at all interested in putting up a repo for this wrapper?

**Comment by meichthys** [comment]
@tejasvi your comment indicated the wrapper was for android but your screenshot shows iOS.
Do you have a repo for this wrapper?
